File: array_equalize.py
#!/usr/bin/python3


def equalizeArray(arr):
    # arr = [3,3,2,1,3]
    numbers = dict()
    for number in arr:
        if numbers.get(number) is None:
            numbers.update({number: 1})
        else:
            numbers[number] += 1
    print(numbers)
    # {1:1, 2:1, 3:3}
    most_occurrences = max(list(numbers.values()))
    kept = False
    amount_to_delete = 0
    for number, amount in numbers.items():
        if amount == most_occurrences and kept is False:
            kept = True
            continue
        amount_to_delete += amount
    print(amount_to_delete)
    return amount_to_delete


equalizeArray([1, 2, 2, 3])  # 2
equalizeArray([1, 2, 2, 3, 4, 5])  # 4
equalizeArray([3, 3, 2, 1, 3])  # 2

equalizeArray([3, 3, 2, 2, 3, 1, 1, 3])  # 4

equalizeArray([1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3])  # 10

File: array_equalize.go
package main

import "fmt"

func equalizeArray(arr []int) int {
	numbers := make(map[int]int)
	for _, number := range arr {
		_, ok := numbers[number]
		if !ok {
			numbers[number] = 1
		} else {
			numbers[number] += 1
		}
	}

	// loop through our constructed enumeration map
	// 	to get the most occurring number
	// could also use math.Max(floats...)
	mostFrequent := 0
	for _, frequency := range numbers {
		if frequency > mostFrequent {
			mostFrequent = frequency
		}
	}

	// simplified - just delete all but the most occurring number
	fmt.Println(len(arr) - mostFrequent)
	return len(arr) - mostFrequent
}

func main() {
	arr1 := []int{1, 2, 2, 3}
	equalizeArray(arr1)

	arr2 := []int{1, 2, 2, 3, 4, 5}
	equalizeArray(arr2)

	arr3 := []int{3, 3, 2, 1, 3}
	equalizeArray(arr3)

	arr4 := []int{3, 3, 2, 2, 3, 1, 1, 3}
	equalizeArray(arr4)

	arr5 := []int{1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3}
	equalizeArray(arr5)
}
